modules/product/commerce_product.module

```diff
@@ -102,7 +103,16 @@ function commerce_product_commerce_product_view(array &$build, EntityInterface $
     /** @var \Drupal\commerce_product\ProductVariationFieldRendererInterface $variation_field_renderer */
     $variation_field_renderer = \Drupal::service('commerce_product.variation_field_renderer');
 
-    $rendered_fields = $variation_field_renderer->renderFields($entity->getDefaultVariation(), $view_mode);
+    // Check if a variation was specified via its canonical URL.
+    // @todo: Core has same problems in regard to translation, follow core's pattern.
```

Comment:
We need to process this at the route level. We need to throw a 404 if variation is not part of the product. And then set the global context if its available, retrieve it here.